### ISSUE TITLE
structure / approach to saving queries in ES

### ISSUE THREAD

**Issue by Analect** [oguennec/angular-elastic-runner#1]
@oguennec 
Thanks for sharing this project publically. I came across it via [angular-elastic-builder](https://github.com/dncrews/angular-elastic-builder/), which you have some involvement in too.

I'm exploring the best approaches to how queries (on top of the query DSL) can be easily generated and then saved within ES .. as your repo is doing.

I can see that the default query [here](https://github.com/oguennec/eean-docker-compose/blob/master/web/src/public/app/querybuilder/queryBuilder.ctrl.js#L51-L81) ... and in `builder` form below ... 

![image](https://cloud.githubusercontent.com/assets/4063815/16724973/1fa346a2-474d-11e6-8189-639356bd6cac.png)

... ends up getting stored on ES as [this](https://gist.github.com/Analect/20332889aa353ff413daca14bf4ed046).  I wanted to get a better understanding of why you might have structured it that way ... and why the `queryObject` contains both `query` and `queryFilter` blocks that appear to be duplicates?

![image](https://cloud.githubusercontent.com/assets/4063815/16724831/7281b1ca-474c-11e6-9bbf-56a340a7485c.png)

I'm assuming the `queryId` and `label` fields are just internal to the app.

With a shift towards bool queries in version 5 (as per blog below), do you anticipate the `angular-elastic-builder` to be extended to handle these types of queries?
https://www.elastic.co/blog/lost-in-translation-boolean-operations-and-filters-in-the-bool-query

Thanks.
Colum

**Comment by oguennec** [comment]
Thanks @Analect  for your interest in this project !

Indeed `queryFilter` which serves as input to angular-elatic-builder is intentionally separated from the `query` block so that we are able to manage queries independently from the `filterQuery`.

Currently `queryFilter` is used to build a filtered query which is deprecated in ES 2.x as you said [filtered query deprecated](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-filtered-query.html) and probably removed in ES 5.x.

My intention is to provide an end-user query object flexible enough to adapt to different ElasticSearch versions or types of query and potentially manage multiple queries in the object. We can drop a `query` but still preserve the `queryFilter` block which is the end-user input.

I like to be able to check that the `queryFilter` can be parsed by angular-elastic-builder on one hand and check the that `query` runs in ElasticSearch as expected.

A next step could be to store ES 5.x and 2.x queries (and their associated sections:  `must`, `should`, `must_not`, `filter`) alongside their ES 1.x equivalent query (and its associated `must`). This could help to test non regression when upgrading to ES 5.x or 2.x for instance. The 1.X ES queries could be ultimately removed.

angular-elastic-builder appears to me as well geared to be adapted to handle ES 5.x/2.x `bool` queries.

Suggestions are welcome.

Olivier

**Comment by Analect** [comment]
@oguennec / Olivier
Thanks for your response.
Can I just just clarify ... the `queryFilter` and the `query`, both contained with the `queryObject` are identical ... it's just that the 'query`additionally wraps the 'queryFilter' along with a`match_all` query ... is this so that the returned matches come with a score, for sorting? ... or is there some other reason?

When you say the `queryFilter` is the end-user input ... do you mean that this is what the end-user builds-up using the bootstrap GUI/widgets? ... which then gets translated as the `queryFilter` block?

What changes might be required in [angular-elastic-builder](https://github.com/dncrews/angular-elastic-builder/) in order to be able to handle these ES5.x/2.x bool queries.  I'm assuming it's not a simple case of changing the subTypes in [GroupDirective.html](https://github.com/dncrews/angular-elastic-builder/blob/master/src/tmpl/GroupDirective.html) from `all`/`any`, to having `must` / `must_not` / `should` / `filter`?

Colum

**Comment by oguennec** [comment]
@Analect / Colum

The match query part of a filtered query is used for scoring, if not specified it default to a match_all query (all docs will get score 1). I chose to explicitly specify the match_all query. An enhancement would be to be able to specify a match query through a builder to score documents ...

By end-user input I mean the object that is being built by angular-elastic-builder and store in `queryFilter`. eean-docker-compose is actually building a query that wraps the `queryFilter`. I could integrate the `queryFilter` in the wrapper query in the code or get rid of the `queryFilter` block in the `queryObject` but I chose to store ES runnable `query` and the `queryFilter` separately in this development phase for the reasons explained above. I expect the `queryObject` structure to evolve as this project progresses.

The angular-elastic-builder question should be posted on dncrews/angular-elastic-builder project. I have not assessed in details how much work an adaptation to handle ES5.x/2.x bool requires yet but I encourage you to give it a try.

Olivier
